### PR TITLE
Instance to Semantic Segmentation Framework

### PR DESCRIPTION
**Pull request by MattsonCam**
This pr implements the changes to perform instance to semantic segmentation using a UNet model. This also includes evaluating training results. The README may be updated in the next pr with results from model training.

### CODE REVIEW

File: callbacks/Callbacks.py [hunk 1]
```diff
@@ -74,36 +74,35 @@ def _log_epoch_metrics(
 
         with torch.no_grad():
             for samples in dataloader:
-                with autocast(enabled=kwargs["use_amp"], device_type=device):
-                    generated_predictions = model(samples["input"])
-                    sigmoid_generated_predictions = generated_predictions.clone()
+                generated_predictions = model(samples["input"])
```
Comment (wli51) [inline callbacks/Callbacks.py:77]: Out of curiosity, it seems like you are switching back from mixed precision to plain fp32/fp16. It is because you simply have more VRAM and you can or did you notice a significant performance penalty with amp?

File: callbacks/utils/SaveEpochSlices.py [hunk 2]
```diff
@@ -18,11 +18,13 @@ class SaveEpochSlices:
     def __init__(
         self,
         image_dataset: torch.utils.data.Dataset,
+        mask_idx_mapping: dict[int, str],
```
Comment (wli51) [inline callbacks/utils/SaveEpochSlices.py:21]: Please correct me if I am wrong: I imagine this is added to handle the multi-mask semantic segmentation scenario and would look some thing like `{0: "nuclei", 1: "cyto", ...}`. 
If that is the case, I wonder if a dictionary mapping makes most sense here for code clarity. If the channel index for different masks are mostly consistent across batches/datasets, having something that is more structured such as a schema would improve readability and make managing definitions easier in the future. If your data is more messy and the flexibility a dictionary offers is necessary, please disregard this comment.

For a schema, I was thinking of something like:
```python
class MaskName(str, Enum):
    NUCLEI = "nuclei"
    CYTO = "cytoplasm"
    MITO = "mitochondria"

@dataclass(frozen=True)
class ChannelSpec:
    name: MaskName             # human-stable identifier
    index: int                 # channel index in tensor
    required: bool = True      # can training proceed if missing?

@dataclass(frozen=True)
class SegmentationSchema:
    version: str               # bump on change
    input_modalities: List[str]  # e.g. ["BF"] or ["DNA","ER",...]
    targets: Dict[MaskName, ChannelSpec]

    def index_for(self, name: MaskName) -> int:
        return self.targets[name].index
``` and we can have some preset realizations of these SegmentationSchema in a registry, importable and directly passed to the SaveEpochSlices.

File: callbacks/utils/SaveWholeSlices.py [hunk 3]
```diff
@@ -201,32 +197,31 @@ def __call__(
...
-                image_path=self.image_dataset[sample_idx]["target_path"],
+                image_path=image_sample["target_path"],
                 image_type="target",
-                image=self.image_dataset[sample_idx]["target"],
+                image=image_sample["target"],
             )
 
             # Only save these images if the segmentation mask isn't black
             # We expect the model to generate black segmentation crops,
             # which will present regardless of weather or not the whole segmented image
```
Comment (wli51) [inline callbacks/utils/SaveWholeSlices.py]:
```suggestion
            # which will present regardless of whether or not the whole segmented image
```
Typo before this PR?

File: callbacks/utils/SaveWholeSlices.py [hunk 4]
```diff
@@ -22,6 +23,7 @@ def __init__(
         image_specs: dict[str, Any],
         stride: tuple[int],
         crop_shape: tuple[int],
+        mask_idx_mapping: dict[int, str],
```
Comment (wli51) [inline callbacks/utils/SaveWholeSlices.py:26]: Same comments as the one in `SaveEpochSlices`, if applicable

File: callbacks/utils/image_padding_specs.py [hunk 5]
```diff
@@ -83,6 +83,7 @@ def compute_patch_mapping(
     )
 
     return (padding_slices, padding_height, padding_width), (
+        slice(None),
```
Comment (wli51) [inline callbacks/utils/image_padding_specs.py:90]: I think your docstring is outdated with the addition of this extra dimension

File: datasets/dataset_00/utils/ImagePreProcessor.py [hunk 6]
```diff
@@ -18,7 +19,8 @@ def __init__(
         target_transform: Optional[callable] = None,
     ):
         self.image_specs = image_specs
-        self.device = device
+        self.device = torch.device(device)
```
Comment (wli51) [inline datasets/dataset_00/utils/ImagePreProcessor.py]: Nitpicking consider also allowing the user to pass in directly a torch.device
Some thing as simple as 
```python
def _to_device(device) -> torch.device:
    return device if isinstance(device, torch.device) else torch.device(device)

def __init__(self, device="cuda",...):
    ...
    self.device = _to_device(device)
    ...
```

File: metrics/BCE.py [hunk 7]
```diff
@@ -14,24 +14,53 @@ class BCE(AbstractMetric):
...
         use_logits: bool = True,
-        reduction: str = "mean",
         device: Union[str, torch.device] = "cuda",
     ):
         super().__init__()
+
+        self.mask_idx_mapping = mask_idx_mapping
+
+        self.mask_weights = (
+            torch.ones(3, device=device) if mask_weights is None else mask_weights
```
Comment (wli51) [inline metrics/BCE.py:28]: I think a saw a lot of magic number 3s which I am assuming is the number of semantic segmentation masks your model is prediction. For the sake of expandability, consider parametrizing this mask number.

File: train.py [hunk 8]
```diff
@@ -65,19 +69,33 @@ def __call__(self, trial):
...
+            is_loss=True,
+            use_logits=True,
+            device=device,
         )
 
         # We do not care about the gradient stability when evaluating performance
+        # This is for recording BCE in mlflow instead of calculating the loss for model updates
         loss_callbacks = BCE(
-            is_loss=True, use_logits=True, reduction="mean", device=device
+            mask_idx_mapping=mask_idx_mapping,
```
Comment (wli51) [inline train.py:83]: Consider making this a parameter or explicitly comment that this class depends on this global variable.

**Review by wli51 — APPROVED** [review]
Great work! PR LGTM!
The only things that stood out to me is the `mask_idx_mapping` and the magic number 3 for adding semantic segmentation dimension. Given that they both exist in various parts of the code and can be immensely helpful for future expandability if made configurable. Still I understand these may take a long time to implement and probably should not be prioritized for now.

File: callbacks/Callbacks.py (hunk 1 above)
Comment (MattsonCam) [reply callbacks/Callbacks.py:77]: It's not, instead of implementing this in callbacks I wrap the trainer call to callbacks with the autocast context manager.

File: callbacks/utils/SaveEpochSlices.py (hunk 2 above)
Comment (MattsonCam) [reply callbacks/utils/SaveEpochSlices.py:21]: It's a good thought, and it has crossed my mind. This approach might change in the future, so I'm just keeping it simple for now.

File: callbacks/utils/SaveWholeSlices.py (hunk 3 above)
Comment (MattsonCam) [reply callbacks/utils/SaveWholeSlices.py]: Good catch, I'll change this. Maybe my head was stuck in the clouds

File: callbacks/utils/image_padding_specs.py (hunk 5 above)
Comment (MattsonCam) [reply callbacks/utils/image_padding_specs.py:90]: Good to know, I'll change it

File: datasets/dataset_00/utils/ImagePreProcessor.py (hunk 6 above)
Comment (MattsonCam) [reply datasets/dataset_00/utils/ImagePreProcessor.py]: Thanks for pointing this out, I thought the torch.device function would handle this. I will change it

File: callbacks/utils/SaveEpochSlices.py (hunk 2 above)
Comment (MattsonCam) [reply callbacks/utils/SaveEpochSlices.py:21]: Also, the mask here doesn't refer to the channel, but the semantic class, which for this application would be (background, cell boundary, and inner cell). Maybe I should add more documentation about this

File: metrics/BCE.py (hunk 7 above)
Comment (MattsonCam) [reply metrics/BCE.py:28]: I think this is generally a good idea, however in this case I don't think the number of semantic classes will change.

File: train.py (hunk 8 above)
Comment (MattsonCam) [reply train.py:83]: Good point, I'll probably just make this a global variable

File: callbacks/utils/SaveWholeSlices.py (hunk 4 above)
Comment (MattsonCam) [reply callbacks/utils/SaveWholeSlices.py:26]: Sounds good, see the corresponding response

**Comment by MattsonCam** [comment]
Thanks @wli51, merging now!
